compiler/rustc_typeck/src/coherence/builtin.rs

```diff
@@ -523,8 +523,14 @@ pub fn coerce_unsized_info<'tcx>(tcx: TyCtxt<'tcx>, impl_did: DefId) -> CoerceUn
                 }
 
                 let (i, a, b) = diff_fields[0];
-                let kind = ty::adjustment::CustomCoerceUnsized::Struct(i);
-                (a, b, coerce_unsized_trait, Some(kind))
+                if Some(def_a.did()) == tcx.lang_items().just_metadata() {
+                    assert_eq!(i, 0, "expected JustMetadata to be coerced in field 0");
+                    let a = substs_a.type_at(0);
+                    let b = substs_b.type_at(0);
+                    (a, b, unsize_trait, CoerceUnsizedKind::JustMetadata)
```

Comment:
I'm still confused why this isn't just doing the same thing as below -- is there an error if you do this?

```suggestion
                    assert_eq!(i, 0, "expected JustMetadata to be coerced in field 0");
                    (a, b, coerce_unsize_trait, CoerceUnsizedKind::JustMetadata)
```

Reply:
that is, I am skeptical of the earlier comment that this would register a `<T as Pointee>::Metadata: CoerceUnsized<<U as Pointee>::Metadata>` obligation. I did a tiny bit of print-debugging and it doesn't seem to be the case.

Reply:
with `unsize_trait`,

```
error[E0277]: the trait bound `<T as Pointee>::Metadata: Unsize<<U as Pointee>::Metadata>` is not satisfied
  --> library\core\src\ops\unsize.rs:75:1
   |
75 | impl<T: ?Sized + Unsize<U>, U: ?Sized> CoerceUnsized<JustMetadata<U>> for JustMetadata<T> {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Unsize<<U as Pointee>::Metadata>` is not implemented for `<T as Pointee>::Metadata`
   |
   = note: all implementations of `Unsize` are provided automatically by the compiler, see <https://doc.rust-lang.org/stable/std/marker/trait.Unsize.html> for more information
help: consider further restricting the associated type
   |
75 | impl<T: ?Sized + Unsize<U>, U: ?Sized> CoerceUnsized<JustMetadata<U>> for JustMetadata<T> where <T as Pointee>::Metadata: Unsize<<U as Pointee>::Metadata> {}
   |                                                                                           ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```

with `coerce_unsize_trait:

```
error[E0277]: the trait bound `<T as Pointee>::Metadata: CoerceUnsized<<U as Pointee>::Metadata>` is not satisfied
  --> library\core\src\ops\unsize.rs:75:1
   |
75 | impl<T: ?Sized + Unsize<U>, U: ?Sized> CoerceUnsized<JustMetadata<U>> for JustMetadata<T> {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `CoerceUnsized<<U as Pointee>::Metadata>` is not implemented for `<T as Pointee>::Metadata`
   |
help: consider further restricting the associated type
   |
75 | impl<T: ?Sized + Unsize<U>, U: ?Sized> CoerceUnsized<JustMetadata<U>> for JustMetadata<T> where <T as Pointee>::Metadata: CoerceUnsized<<U as Pointee>::Metadata> {}
   |                                                                                           +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```

`a`, `b` here are the types of the field in order to require `CoerceUnsize` for the fields that are changed by the type change for struct coercions. For `JustMetadata`, we want `T: Unsize<U>`, not `Field<T>: CoerceUnsize<Field<U>>`.

Reply:
you're right, I read through the code more thorougly and understand now. sorry for the confusion!

Reply:
side-note: maybe we should lift this into a different branch in the match, and skip all the other checking that happens here, e.g.

```rust
 (&ty::Adt(def_a, substs_a), &ty::Adt(def_b, substs_b))
                if Some(def_a) == tcx.lang_items().just_metadata() && Some(def_b) == tcx.lang_items().just_metadata() =>
    (substs_a.type_at(0), substs_b.type_at(0), unsized_trait, CoerceUnsizedKind::JustMetadata),
```

Reply:
I'd be more comfortable doing so if we add validation that the lang item only has the one field. As is, this here is what ends up validating that `JustMetadata` has the correct shape.

Reply:
I think you could do that, perhaps emitting a `bug!()` if it has >1 field. You could even assert that it's equal to the projection type `<$SUBST_0 as Pointee>::Metadata` (where `$SUBST_0` is the 0th subst type). 

But up to you.

Reply:
I went ahead and pushed this simplification; I don't know how much we care about nice/usable errors for `#[lang]` items are defined incorrectly. ~~...Also I went looking and couldn't find the place to add the wf check; `rustc_typeck::check::wfcheck::check_type_defn` seems like a candidate but I both don't know how to `LocalDefId == DefId` nor how to actually add the `field == <T as Pointee>::Metadata` obligation.